lib/windows/app/actions/deviceActions.js

```diff
@@ -0,0 +1,313 @@
+
+/**
+ * Indicates that devices have been detected. This is triggered by default at
+ * startup, and whenever a device is attached/detached. The app can configure
+ * which devices to look for by providing a `config.selectorTraits` property.
+ *
+ * @param {Array} devices Array of all attached devices, ref. nrf-device-lister.
+ */
+export const DEVICES_DETECTED = 'DEVICES_DETECTED';
+
+let deviceLister;
+
+// Defined when user input is required during device setup. When input is
+// received from the user, this callback is invoked with the confirmation
+// (Boolean) or choice (String) that the user provided as input.
+let deviceSetupCallback;
+
+function deviceSelectedAction(device) {
+    return {
+        type: DEVICE_SELECTED,
+        device,
+    };
+}
+
+function deviceSetupErrorAction(device, error) {
+    return {
+        type: DEVICE_SETUP_ERROR,
+        device,
+        error,
+    };
+}
+
+function deviceSetupCompleteAction(device) {
+    return {
+        type: DEVICE_SETUP_COMPLETE,
+        device,
+    };
+}
+
+function devicesDetectedAction(devices) {
+    return {
+        type: DEVICES_DETECTED,
+        devices,
+    };
+}
+
+function deviceSetupInputRequiredAction(message, choices) {
+    return {
+        type: DEVICE_SETUP_INPUT_REQUIRED,
+        message,
+        choices,
+    };
+}
+
+function deviceSetupInputReceivedAction(input) {
+    return {
+        type: DEVICE_SETUP_INPUT_RECEIVED,
+        input,
+    };
+}
+
+function logDeviceListerError(error) {
+    if (error.usb) {
+        const usbAddr = `${error.usb.busNumber}.${error.usb.deviceAddress}`;
+
+        let message = `Error while probing usb device at bus.address ${usbAddr}: ${error.message}. `;
+        if (process.platform === 'linux') {
+            message += 'Please check your udev rules concerning permissions for USB devices, see ' +
+                'https://github.com/NordicSemiconductor/nrf-udev';
+        } else if (process.platform === 'win32') {
+            message += 'Please check that a libusb-compatible kernel driver is bound to this device.';
+        }
+
+        logger.error(message);
+    } else {
+        logger.error(`Error while probing devices: ${error.message}`);
+    }
+}
+
+/**
+ * Deselects the currently selected device.
+ *
+ * @returns {Object} Action object that can be passed to redux dispatch.
+ */
+export function deselectDevice() {
+    return {
+        type: DEVICE_DESELECTED,
+    };
+}
+
+/**
+ * Starts watching for devices with the given traits. See the nrf-device-lister
+ * library for available traits. Whenever devices are attached/detached, this
+ * will dispatch DEVICES_DETECTED with a complete list of attached devices.
+ *
+ * @returns {function(*)} Function that can be passed to redux dispatch.
+ */
+export function startWatchingDevices() {
+    const config = getAppConfig();
+
+    return (dispatch, getState) => {
+        deviceLister = new DeviceLister(config.selectorTraits);
```

Comment:
No need to instantiate a new `DeviceLister` if the instance exists already.
This will probably leak memory (and prevent the process from closing cleanly) if `startWatchingDevices` is called several times without the corresponding `stopWatchingDevices()`.

Reply:
Good point. Will add a check here, so that it does not recreate the instance if it does not already exist.

The `let deviceLister` declaration is evaluated before `loadApp()` from util/apps.js has been called, so at that point we don't have `config.selectorTraits` yet.